poa-token/src/contract.rs

```diff
@@ -56,6 +78,115 @@ impl Contract {
+
+        Self {
+            token: old_state.token,
+            metadata: old_state.metadata,
+            wrapped_token: None,
+        }
+    }
+
+    #[only(self, owner)]
+    pub fn clear_wrapped_token_account_id(&mut self) {
+        self.wrapped_token = None;
+    }
+
+    #[only(self, owner)]
+    pub fn set_wrapped_token_account_id(&mut self, token_account_id: AccountId) -> Promise {
+        if self.wrapped_token.is_some() {
+            env::panic_str("Wrapped token is already set")
+        }
+
+        ext_ft_core::ext(token_account_id.clone())
+            .ft_balance_of(CURRENT_ACCOUNT_ID.clone())
+            .then(
+                Self::ext(CURRENT_ACCOUNT_ID.clone())
+                    .with_static_gas(DO_WRAP_TOKEN_GAS)
+                    .do_set_wrapped_token_account_id(token_account_id),
+            )
+    }
+
+    #[only(self, owner)]
+    pub fn do_set_wrapped_token_account_id(&mut self, token_account_id: AccountId) {
+        let parsed_balance = match env::promise_result(0) {
+            near_sdk::PromiseResult::Successful(balance_bytes) => {
+                if let Ok(balance) = near_sdk::serde_json::from_slice::<U128>(&balance_bytes) {
+                    balance
+                } else {
+                    env::panic_str(&format!(
+                        "Setting token id {token_account_id} for contract {} failed due to bad balance bytes",
+                        &*CURRENT_ACCOUNT_ID
+                    ))
+                }
+            }
+            near_sdk::PromiseResult::Failed => env::panic_str(
+                "Setting token id {token_account_id} for contract {} failed due to failed promise",
+            ),
+        };
+
+        let self_total_supply = self.ft_total_supply();
+        require!(parsed_balance >= self_total_supply);
+
+        self.wrapped_token = Some(token_account_id);
+    }
+
+    #[only(self, owner)]
+    pub fn ft_resolve_unwrap(&mut self, account_id: &AccountId, amount: u128, is_call: bool) {
+        let unused_amount = match env::promise_result(0) {
+            near_sdk::PromiseResult::Successful(value) => {
+                if let Ok(unused_amount) = near_sdk::serde_json::from_slice::<U128>(&value) {
+                    std::cmp::min(amount, unused_amount.0)
+                } else {
+                    amount
+                }
+            }
+            near_sdk::PromiseResult::Failed => {
+                // TODO: understand this logic
+                if is_call {
+                    // do not refund on failed `ft_transfer_call` due to
+                    // NEP-141 vulnerability: `ft_resolve_transfer` fails to
+                    // read result of `ft_on_transfer` due to insufficient gas
+                    amount
+                } else {
+                    0
+                }
+            }
+        };
+
+        self.token.internal_deposit(account_id, unused_amount);
+    }
+}
+
+impl Contract {
+    fn do_unwrap_and_transfer(
```

Comment:
There is no preprocessing step here, so `do_` prefix is not needed
```suggestion
    fn unwrap_and_transfer(
```